--- a/src/Data/OpenApi/Internal/Schema.hs
+++ b/src/Data/OpenApi/Internal/Schema.hs
@@ -625 +625,14 @@ instance (Typeable (Fixed a), HasResolution a) => ToSchema (Fixed a) where decla
-  declareNamedSchema _ = declareNamedSchema (Proxy :: Proxy a)
+  declareNamedSchema _ = do
+    namedSchema <- declareNamedSchema (Proxy :: Proxy a)
+    pure
+      namedSchema
+      { _namedSchemaName = fmap ("Maybe_" <>) (_namedSchemaName namedSchema)
+      , _namedSchemaSchema =
+          _namedSchemaSchema namedSchema
+            & nullable ?~ True
+      }
+type NestedMaybeError =
+  Text "Nested " :<>: ShowType Maybe :<>: Text "s are not supported." :$$:
+  Text "OpenAPI 3 only supports a single level of nullability, so it can't distinguish between Nothing and Just Nothing."
+instance {-# OVERLAPPING #-} (TypeError NestedMaybeError, Typeable a) => ToSchema (Maybe (Maybe a)) where
+  declareNamedSchema _ = undefined
@@ -1020,2 +1033,9 @@ instance {-# OVERLAPPABLE #-} (Selector s, GToSchema f) => GToSchema (S1 s f) w
 instance {-# OVERLAPPING #-} ToSchema c => GToSchema (K1 i (Maybe c)) where
-  gdeclareNamedSchema _ _ _ = declareNamedSchema (Proxy :: Proxy c)
+  gdeclareNamedSchema _ _ _ = unnamed . nullableSchemaRef <$> declareSchemaRef (Proxy :: Proxy c)
+nullableSchemaRef :: Referenced Schema -> Schema
+nullableSchemaRef (Inline schema) =
+  schema & nullable ?~ True
+nullableSchemaRef schemaRef =
+  mempty
+    & allOf ?~ [schemaRef]
+    & nullable ?~ True
